### PR TITLE
[Dynamo] Automatically in-graph traceable tensor subclass ctors

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/114389

Previously, dynamo would attempt to trace through the `__init__` of traceable tensor subclasses, since their constructors are AOT dispatcher traceable by definition, dynamo should automatically put these in the graph like we do for any other tensors. Not doing this is difficult because dynamo would need to apply mutations post tensor subclass creation in the graph. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135151

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/utils/_python_dispatch.py [hunk 1]
```diff
@@ -391,6 +391,11 @@ def is_traceable_wrapper_subclass(t: object) -> TypeGuard[TensorWithFlatten]:
         and hasattr(t, "__tensor_unflatten__")
     )
 
+def is_traceable_wrapper_subclass_type(t: Type) -> TypeGuard[Type[TensorWithFlatten]]:
```
Comment (bdhirsh) [inline torch/utils/_python_dispatch.py:394]: output of this fn should be `-> bool:` right?

Comment (bdhirsh) [reply torch/utils/_python_dispatch.py:394]: (oh same with the one above but I guess that is old)

Comment (bdhirsh) [reply torch/utils/_python_dispatch.py:394]: also nit: can you just update `is_traceable_wrapper_subclass` to `return is_traceable_wrapper_subclass_type(type(x))` to dedup the code?

Comment (mlazos) [reply torch/utils/_python_dispatch.py:394]: Oh I thought TypeGuard is just a special bool, right?

File: test/dynamo/test_subclasses.py [hunk 2]
```diff
@@ -1427,6 +1428,62 @@ def __metadata_guard__(self, x, y):
             lambda: torch.compile(lambda x: x * x)(x),
         )
 
+    def test_subclass_constructor_proxying(self):
+        class SubclassTensor(torch.Tensor):
```
[the remark this answers is not in the document]
Comment (mlazos) [reply test/dynamo/test_subclasses.py:1432]: Yup, once the other PR merges I'll test both these cases. Good idea!

**Review by bdhirsh — APPROVED** [review]
lgtm minus nits!

**Pull request closed without merging**